veps/vep-4.md

```diff
@@ -0,0 +1,342 @@
+W1031 13:17:56.497648   88580 servenv.go:270] Use of single-dash long flags is deprecated and will be removed in 14.0. Please use --cell instead
+```
+
+The second issue with dropping in `pflag` is that it supports mixing of positional and flag arguments.
+**This change will break subcommands, such as those in `vtctl` and `mysqlctl`.**
+
+To illustrate the problem, consider the command `./mycmd -a 1 -b 2 X -c 3 -d 4`.
+With `flag`, parsing this will result in two flags, `a` and `b`, `flag.NArg()` will return 5 and `flag.Args()` will return the slice `["X", "-c", "3", "-d", "4"]` ([c.f.](https://play.golang.org/p/sO4UlQPGg4l)).
+With `pflag`, all 4 flags will be parsed, `pflag.NArg()` will return 1, and `pflag.Args()` will return the slice `["X"]`.
+
+In order to match this behavior, users should update their commands to insert a double-dash (`--`) before the first positional argument.
+This causes `pflag` to stop looking for flags and treat everything after as a positional argument (matching what `flag` does even without the double-dash), and has no effect on the parsing behavior of `flag` ([c.f.](https://play.golang.org/p/76FrFrLFOMT)).
+
+As with the single-dash long flag issue, we can make code changes similar to the `parse` snippet above to warn users and target v14 for switching, or we can announce and move to `pflag` immediately.
+
+### Phase 2: `pflag` adoption and migration
+
+After the preparatory work we did in phase 1, dropping in `pflag` is fairly straightforward.
+For example, `servenv` has two functions, `ParseFlags` and `ParseFlagsWithArgs`.
+Taking the first as an example, we would write:
+
+```go
+import (
+    goflag "flag"
+    flag "github.com/spf13/pflag" // minimize code changes
+)
+func ParseFlags(cmd string) {
+    flag.CommandLine.AddGoFlagSet(goflag.CommandLine) // (TODO) Remove after all other packages add flags directly to pflag.
+	flag.Parse()
+
+	if *Version {
+		AppVersion.Print()
+		os.Exit(0)
+	}
+
+	args := flag.Args()
+	if len(args) > 0 {
+		flag.Usage()
+		log.Exitf("%s doesn't take any positional arguments, got '%s'", cmd, strings.Join(args, " "))
+	}
+}
+```
+
+Now, other packages which declare flags can begin moving to `pflag`.
+Once all packages are adding their flags directly to the `pflag.CommandLine` flagset, we can remove the "TODO" interop line.
+Now, other packages can gradually move over to `pflag`, and once everything is adding their flags directly to the `pflag` flagset, we can remove the interop line.
+
+Note that there are other places that call `flag.Parse` directly (most in `go/test`, some in `go/cmd`) that will need the same treatment.
+In addition, places with subcommands will need to update callsites to take a `pflag.FlagSet`, but the changes there should be fairly minimal.
+For example, modulo running `goimports`, [this][mysqlctl_pflag_diff] is the full diff required to compile `mysqlctl`.
```

Comment:
Should that link be to a gist or something like that? Also I couldn't really parse this sentence.

Reply:
It does (easier to see in the Rich Diff view); the gist it links to is defined allll the way down [here](https://github.com/vitessio/enhancements/pull/8/files#diff-c8fb26a92eec12c34cf779d76b5b05f9f9ac06c04e78d61584ff897428748018R323) with the rest of the links.

I can remove the "modulo running `goimports`" bit (I'm guessing that's where this gets wordy/confusing?); the only reason I added it is it's _technically_ not the actual diff because if you run `goimports` it (correctly) moves the `pflag` import down to the 3rd-party section which makes the diff slightly larger/less clear to read. (To be clear: not suggesting we ship a diff that breaks goimports for this work, only that doctoring the diff a bit makes the clarity of the change easier to see for the purpose of the RFC proposal)